### PR TITLE
Adds TTS to all speech

### PR DESCRIPTION
**Pull request by ynot01**
# Document the changes in your pull request

aeiou is real

##

Adds TTS similar to Moonbase Alpha to all speech, mechanical and organic

--> **You will not hear the TTS unless you have the preference on** <--

People will probably not meme as much because not everyone has TTS on and thus it will be less funny

**You will be able to select your voice in character preferences or leave it random**

Ghosts cannot TTS

You will not hear TTS from someone who is speaking a language you do not understand

TTS audio is local only- you will not hear TTS over the radio... unless? 👉👈

Some have presented the idea of having this as a pref toggle, I can do that if it will get it merged

Generated audio files are stored for 10 seconds in order to reduce performance cost on spammy messages (i.e. every newscaster triggering at once as they do) and are then deleted

Leftover messages (messages sent in the last 10 seconds of a round, probably) are deleted on the initialization of a new round, so disk space does not accumulate and cause problems in the future

Sanitation (`san_message`) needs review by someone who knows at least basic security (@alexkar598), I covered what I thought could be dangerous in cmd but I'm by no means an expert

Uses say.exe and libs from [80speak's CLI](https://github.com/connornishijima/80speak) which is licensed under [GNU GPL v3](https://www.gnu.org/licenses/gpl-3.0.html) (this is added to README.md)

https://user-images.githubusercontent.com/28408322/178133134-a82611f7-9017-45b0-84fb-b3d6de7f243f.mp4

# Changelog

:cl:  
rscadd: Added TTS
/:cl:

### CODE REVIEW

**Comment by stwordcitizen** [comment]
Reason I am against this is because there's already so much audio going on at once in a round like the footsteps radio sounds and ambient noise, none of those sound good to begin with and adding some microsoft sam voice that can't be toggled off as far as I can tell would make it worse. Also I don't need to have constant audio input from some stuttering person who talks to much.

**Comment by SomeguyManperson** [comment]
hey you know how we added a cooldown on emotes specifically so people couldn't spam deathgasp as xenos

yeah let's do the exact opposite

**Review by alexkar598 — CHANGES_REQUESTED** [review]
Do not merge or testmerge until security review.

Furthermore, you need to make sure this is cross platform, we do not run windows and we arent about to run wine in prod.

**Comment by maxion12345** [comment]
> Reason I am against this is because there's already so much audio going on at once in a round like the footsteps radio sounds and ambient noise, none of those sound good to begin with and adding some microsoft sam voice that can't be toggled off as far as I can tell would make it worse. Also I don't need to have constant audio input from some stuttering person who talks to much.

reason I am against you going against this is that 
A) comedic relief
B) actually might help in stressful situations with lots of people
C) it'll be a perf so if you don't want it you don't need it (although IMO it'll be like playing fullscreen while others play widescreen)


should testmerge it anyways just so we can have some rounds of funny tts

**Comment by Cartlord** [comment]
I do not like the idea of having the pref be added later - it seems entirely likely, to me, that this will be added, and then nobody will get around to actually adding the pref.

**Comment by ynot01** [comment]
linux solution coming tmrw

**Comment by Bibby0110** [comment]
What’s the performance impact?

**Comment by ynot01** [comment]
> I do not like the idea of having the pref be added later - it seems entirely likely, to me, that this will be added, and then nobody will get around to actually adding the pref.

if maints tell me they want it on pref it would be done on this PR unless they merge it before I can do so

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
Absolutely going to have to be a preference

**Comment by maxion12345** [comment]
Pref would be a requirement but it'd be an interesting feature to add to the mix. I'd just have it start disabled by default, as it's not really a 'required' thing that'll give any advantage (like widescreen and runechat). It'll just be for the ones who want to have a giggle and maybe be useful in some situations

**Comment by N3D6** [comment]
This would be dope as fuck if it was a preference and not mandatory

**Comment by ynot01** [comment]
linux ver is untested as i do not have a linux machine or VM at the ready, i expect it to work but after several hours of frustration my linux VM bricked so i cant test it myself

im going to continue development on other things (prefs, more voices) while that sits there

**Comment by ynot01** [comment]
> What’s the performance impact?

From testing, minimal; even extremely long strings are done instantly

**Comment by ynot01** [comment]
prefs added:
- you can now select which voice to start with, of 8 voices; 4 masculine 4 femine or random
- you now have to enable TTS in settings dropdown (below typing indicators) to hear TTS

**Comment by ynot01** [comment]
everything is tested and in order except for the linux cmd which i cannot test locally but it probably works

File: code/modules/mob/dead/new_player/new_player.dm
Comment (tattax) [inline code/modules/mob/dead/new_player/new_player.dm]:
```suggestion
		switch(gender)
			if(MALE)
				H.tts = pick("Masc1", "Masc2", "Masc3", "Masc4")
			if(FEMALE)
				H.tts = pick("Fem1", "Fem2", "Fem3", "Fem4")
			else
				H.tts = pick("Masc1", "Masc2", "Masc3", "Masc4", "Fem1", "Fem2", "Fem3", "Fem4")
```
//////////////////////////////////////////////////

```suggestion
		switch(gender)
			if(MALE)
				H.tts = pick("Masc1", "Masc2", "Masc3", "Masc4")
			if(FEMALE)
				H.tts = pick("Fem1", "Fem2", "Fem3", "Fem4")
			if(PLURAL)
				H.tts = pick("Masc1", "Masc2", "Masc3", "Masc4", "Fem1", "Fem2", "Fem3", "Fem4")
```

**Comment by Bibby0110** [comment]
> > What’s the performance impact?
> 
> From testing, minimal; even extremely long strings are done instantly

please find some objective way to measure the performance so we can see if it's feasible on shifts of 60+ people, with some objects also using say()

File: code/__DEFINES/{yogs_defines}/preferences.dm
Comment (Bibby0110) [inline code/__DEFINES/{yogs_defines}/preferences.dm]: make sure this works and doesn't overflow

ALSO

probably needs to be off by default

File: code/game/say.dm
Comment (Bibby0110) [inline code/game/say.dm:42]: probably could use a better check, like invisibility or similar (would allow for admin funny and less edgecases)

File: code/game/say.dm
Comment (Bibby0110) [inline code/game/say.dm]: observers don't get to witness the glory of TTS?

File: code/game/say.dm
Comment (Bibby0110) [inline code/game/say.dm]: shouldn't it be 
```suggestion
			if(MB.client && (MB.client.prefs.chat_toggles & HEAR_TTS) && MB.has_language(message_language))
```
idk don't quiz me on bitwise operations

File: code/modules/client/preferences_savefile.dm
Comment (Bibby0110) [inline code/modules/client/preferences_savefile.dm]: wouldn't this turn on HEAR_TTS

File: code/modules/client/preferences.dm
Comment (Bibby0110) [inline code/modules/client/preferences.dm]: global list

File: code/modules/mob/dead/new_player/new_player.dm
Comment (Bibby0110) [inline code/modules/mob/dead/new_player/new_player.dm]: use global list

File: code/modules/mob/living/carbon/human/human.dm
Comment (Bibby0110) [inline code/modules/mob/living/carbon/human/human.dm]: tbh they should also probably be defines

**Review by Bibby0110 — CHANGES_REQUESTED** [review]

File: code/__DEFINES/{yogs_defines}/preferences.dm
Comment (ynot01) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: is off by default

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm]: nah, hard to explain but basically as it stands the pref (labelled Hear TTS) starts off (unchecked) and this is just it works; ive done extensive testing regarding this

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm]: shrug

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm:42]: ghosts being un-invisible doesnt mean you should hear what they say

File: code/modules/client/preferences_savefile.dm
Comment (ynot01) [reply code/modules/client/preferences_savefile.dm]: i don't think it does? ill test later

**Comment by ynot01** [comment]
changes & cooldown coming soon

File: code/game/say.dm
Comment (tattax) [inline code/game/say.dm]:
```suggestion
			if(MB?.client?.prefs.chat_toggles & ~HEAR_TTS && MB.has_language(message_language))
```

Comment (tattax) [reply code/game/say.dm]: something something this should work

File: code/modules/client/preferences_savefile.dm
Comment (Bibby0110) [reply code/modules/client/preferences_savefile.dm]: from DM reference:
usr.sight |= BLIND // turn on the blind bit

turns on if not on, if already on does nothing

File: code/game/say.dm
Comment (Bibby0110) [reply code/game/say.dm]: the way the code is written it will be true when HEAR_TTS isn't checked

File: code/game/say.dm
Comment (Bibby0110) [reply code/game/say.dm:42]: why not?

File: code/__DEFINES/{yogs_defines}/preferences.dm
Comment (Bibby0110) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: this is the list of toggles turned on by default

**Comment by JamieD1** [comment]
@alexkar598 Can you do a sec review if you can

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm:58]: why did you negate `HEAR_TTS`?

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm:42]: wizard rounds, spectral blade, you can't hear what those ghosts have to say and it would be a gateway to meta info

Comment (Bibby0110) [reply code/game/say.dm:42]: That’s the point

Comment (ynot01) [reply code/game/say.dm:42]: what

**Comment by ynot01** [comment]
> please find some objective way to measure the performance so we can see if it's feasible on shifts of 60+ people, with some objects also using say()

running the program x times to generate x unique files, with the TTS saying "[number], hello, world" where `number` is the file's chronological place

lipsum tests run the entire first paragraph of [lipsum](https://www.lipsum.com/) in place of "hello, world"

1 file test avg: 183080 μsec, or 183.08 msec
100 files test avg: 187755 μsec, or 187.755 msec
1000 files test avg: 182203 μsec, or 182.203 msec
1 file, lipsum avg: 182814 μsec, or 182.814 msec
50 files, lipsum avg: 192013 μsec, or 192.013 msec

also note that ingame, duplicate messages are re-used instead of re-generated

File: code/__DEFINES/{yogs_defines}/preferences.dm
Comment (ynot01) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: in latest commit [tatax broke my bitflags 0e83962](https://github.com/yogstation13/Yogstation/pull/14775/commits/0e83962ca73176816c8b4e32156aec10a6def64e) i deleted my player save and started fresh and it works (starts off, needs to be toggled on to hear TTS)

Comment (ynot01) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: running a quick test with it removed

Comment (ynot01) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: ok i think i know whats wrong

Comment (ynot01) [reply code/__DEFINES/{yogs_defines}/preferences.dm]: fixed in last commit

File: code/modules/client/preferences.dm
Comment (ynot01) [reply code/modules/client/preferences.dm]: global list now enums these

Comment (ynot01) [reply code/modules/client/preferences.dm]: switched to global list in latest commit

File: code/game/say.dm
Comment (alexkar598) [inline code/game/say.dm]: I want a whitelist on this, not a blacklist

File: code/modules/mob/living/say.dm
Comment (alexkar598) [inline code/modules/mob/living/say.dm]: voice needs to be sanitized

File: code/modules/mob/living/say.dm
Comment (alexkar598) [inline code/modules/mob/living/say.dm]: Whitelist not blacklist again

**Review by alexkar598 — CHANGES_REQUESTED** [review]

**Review by alexkar598 — CHANGES_REQUESTED** [review]
I've played with the linux say.exe for a bit and so far all I can say is that I'm not impressed, I got it to crash several times and it being a monsterous C program has me worried for the overall security of the binary itself.

On top of that, there is no clear way on how we are meant to compile this and I'm not a fan of random binaries being put in the repo.

File: code/modules/mob/living/say.dm
Comment (ynot01) [reply code/modules/mob/living/say.dm]: done in latest commit [b344871](https://github.com/yogstation13/Yogstation/pull/14775/commits/b344871def2bf9859ce5062d67ae0ba97af9abf7)

File: code/game/say.dm
Comment (ynot01) [reply code/game/say.dm]: done in latest commit [b344871](https://github.com/yogstation13/Yogstation/pull/14775/commits/b344871def2bf9859ce5062d67ae0ba97af9abf7)

File: code/modules/mob/living/say.dm
Comment (ynot01) [reply code/modules/mob/living/say.dm]: done in latest commit [b34f28a](https://github.com/yogstation13/Yogstation/pull/14775/commits/b34f28aedf074483f41be7fb49d08cd3d897d737)

File: code/modules/mob/living/say.dm
Comment (alexkar598) [inline code/modules/mob/living/say.dm]: Voice is still not sanitizied:
- tts_enum can be controlled by admins
- H.tts can be controlled by admins

**Review by alexkar598 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/say.dm
Comment (ynot01) [reply code/modules/mob/living/say.dm]: resolved in [239c0e0](https://github.com/yogstation13/Yogstation/pull/14775/commits/239c0e065fc83f90a95cc36c4edce8d78ba66c3f)

File: code/__HELPERS/text.dm
Comment (alexkar598) [inline code/__HELPERS/text.dm:327]: These can be controlled by attackers

File: code/__HELPERS/text.dm
Comment (alexkar598) [inline code/__HELPERS/text.dm:329]: So can these

**Review by alexkar598 — CHANGES_REQUESTED** [review]
I've done a tiny bit of research and it seems that DECTALK is still under copyright. It would be unwise to use a proprietary binary we didnt pay for which is still under copyright.

**Comment by ynot01** [comment]
DECtalk was published in 1984 by a company called Digital Equipment Corporation which was dissolved in 1998 when it was acquired by a company called Compaq, which dissolved in 2002 when it was acquired by Hewlett-Packard(HP).
During its time under Compaq, the DECtalk IP was sold to Force Computers in 2000.
In late 2001, it was sold again, to Fonix Speech, which in 2011 rebranded itself to SpeechFX.
SpeechFX offered DECtalk alongside its reworked FonixTalk and RealVoice SDKs, until its website was seemingly taken down around 2021 and now redirects to a commonly reported virus website

[The organization I acquired these binaries from](https://github.com/dectalk)

I am not a lawyer, but through my research, merging & hosting this could potentially violate the intellectual rights of SpeechFX, Inc.

**Comment by ynot01** [comment]
If anyone finds an open license TTS, this branch could easily be re-opened supporting that one instead

**Pull request closed without merging**
